app/models/miq_queue.rb

```diff
@@ -376,6 +385,29 @@ def self.put_or_update(find_options)
     msg
   end
 
+  def self.update_with_jsonb_data(msg, jsonb_targets, save_options)
+    connection = ActiveRecord::Base.connection
+
+    jsonb_set_query = jsonb_targets.map do |target|
+      target_data = target.second # Stored in format [target_class, target_data]
+      "jsonb_data = jsonb_set(jsonb_data, '{#{target_data[:association]}___#{target_data[:manager_ref]}}', '#{target.to_json}')"
+    end.join(", ")
+
+    options_set_query = save_options.map do |key, value|
+      next if %i(args).include?(key) # args are stored the old way
+
+      "#{connection.quote_column_name(key)} = #{connection.quote(value)}"
+    end.compact.join(", ")
+
+    query = <<-SQL
+      UPDATE miq_queue
+      SET #{jsonb_set_query}, #{options_set_query}
+      WHERE miq_queue.id = #{msg.id}
```

Comment:
I *really* don't like us building queries, especially for MiqQueue.

Reply:
me neither, but Rails doesn't seem to support the jsonb_set

```"jsonb_data = jsonb_set(jsonb_data, '{#{target_data[:association]}___#{target_data[:manager_ref]}}', '#{target.to_json}')"```

Reply:
You can still do 90% of this with ActiveRecord though:

```ruby
def self.update_with_jsonb_data(msg, jsonb_targets, save_options)
  jsonb_set_clause = jsonb_targets.map do |target|
    target_data = target.second # Stored in format [target_class, target_data]
    "jsonb_data = jsonb_set(jsonb_data, '{#{target_data[:association]}___#{target_data[:manager_ref]}}', '#{target.to_json}')"
  end.join(", ")

  set_clause = MiqQueue.send(:sanitize_sql_for_assignment, save_options.except(:args))

  MiqQueue.where(:id => msg.id).update_all("#{set_clause}, #{json_set_clause}")
end
```

Reply:
ah, nice